src/odemis/driver/orsay.py

```diff
@@ -1389,3 +1440,1097 @@ def terminate(self):
+            return
+
+        eState = ""
+
+        heater_state = self._hvps.HeaterState.Actual
+        if heater_state == HEATER_ERROR:  # in case of heater error
+            eState += "FIB source forced to shut down"
+
+        if eState == "":
+            self.state._set_value(model.ST_RUNNING, force_write=True)
+        else:
+            self.state._set_value(HwError(eState), force_write=True)
+
+    def _updateHeater(self, parameter=None, attributeName="Actual"):
+        """
+        parameter (Orsay Parameter): the parameter on the Orsay server to use to update the VA
+        attributeName (str): the name of the attribute of parameter which was changed
+
+        Reads if the FIB source heater is on from the Orsay server and saves it in the heater VA
+        """
+        if parameter is None:
+            parameter = self._hvps.HeaterState
+        if parameter is not self._hvps.HeaterState:
+            raise ValueError("Incorrect parameter passed to _updateHeater. Parameter should be "
+                             "datamodel.HVPSFloatingIon.HeaterState. Parameter passed is %s"
+                             % parameter.Name)
+        if attributeName != "Actual":
+            return
+        heater_state = self._hvps.HeaterState.Actual
+        new_value = False
+        logging.debug("FIB source heater state is: %s." % heater_state)
+        if heater_state in (HEATER_ON, HEATER_RISING, HEATER_FALLING):  # alternative values: HEATER_OFF, HEATER_ERROR
+            new_value = True
+        self.heater._value = new_value  # to not call the setter
+        self.heater.notify(new_value)
+
+    def _changeHeater(self, goal):
+        """
+        goal (bool): goal state of the heater: (True: "ON", False: "OFF")
+        return (bool): goal state of the heater as set to the server: (True: "ON", False: "OFF")
+
+        Turns on the FIB source heater on the Orsay server if argument goal is True. Turns it off otherwise.
+        """
+        logging.debug("Setting FIB source heater to %s." % (HEATER_ON if goal else HEATER_OFF))
+        self._hvps.HeaterState.Target = HEATER_ON if goal else HEATER_OFF
+        return self._hvps.HeaterState.Target == HEATER_ON
+
+    def terminate(self):
+        """
+        Called when Odemis is closed
+        """
+        if self._hvps is not None:
+            for obj_name in self._connectorList:
+                getattr(self, obj_name).disconnect()
+            self._connectorList = []
+            self._hvps = None
+            self._ionColumn = None
+
+
+class FIBBeam(model.HwComponent):
+    """
+    Represents the beam of the Focused Ion Beam (FIB) from Orsay Physics. It contains many beam optics settings.
+    """
+
+    def __init__(self, name, role, parent, **kwargs):
+        """
+        Defines the following VA's and links them to the callbacks from the Orsay server:
+        • blanker: VAEnumerated, choices={True: "blanking", False: "no blanking", None: "imaging"}
```

Comment:
Fair point. Especially since, if the range is defined on the Orsay software, the range specified on the VA will be overwritten by the range defined on the Orsay software. So it might not be correct anyway.

Reply:
Btw, because of the ridiculous amount of parameters to control, we wanted to use something like the Connector class, to prevent repeating a LOT of similar code. And this part of the driver does not actually offer much functionality yet. The Detector and Scanner classes will offer the actual functionality, like imaging, and use the FIBBeam, FIBSource and FIBDevice for communication with the Orsay software.